### PR TITLE
[MDB IGNORE] Shuttle engine code improvement and fixes

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
This is a remake of https://github.com/tgstation/tgstation/pull/69362 because github desktop has fucked over my branch 

## About The Pull Request

I hated how shuttles worked while I was working with them on a downstream, so instead I gave up and started refactoring shit here.

My original goal was solely just to make engines handle being added to lists themselves, but it turned into this instead.

* Makes use of ``as anything`` in many places
* Renames many vars, including shuttle 'id' var to 'shuttle_id' and engine 'state' to 'engine_state', so it is easier to search for them and manage them at once.
* Engines now weakref their attached ship, and disconnect when unwrenched from it, allowing engines to do more things to their shuttle without constantly checking for being inside of one.
* Removed shuttles constantly checking for how many engines they are, the engine will add themselves to the list instead.
* Adds comments to port_id and shuttle_id to help understand what is needed with these two vars, because currently admins can only replace Emergency shuttles and Ferries, the rest are broken entirely, because people named them differently.
* Adds a ``mapload`` arg to ``connect_to_shuttle()`` for when a shuttle is first created and is syncing machines to it, since in some occasions, stuff may be connected to shuttles afterwards.

In this PR, I fixed Cargo, Arrivals, and Pirate ships. Emergency and Whiteship ones are already functional, the rest I am leaving because:
1-lazy
2-they arent really that important
3-some shouldn't even be replacing eachother? fugitive hunters one seems a little odd, and I don't really want to get too deep into that, I wanted to focus more on the ship parts itself.

## Why It's Good For The Game

I hate the current state of shuttles, especially when they are straight-up broken. Hopefully I can make it slightly better and make it easier to work with in the future, while fixing a few things on the way. Like my vault console PR, this is somewhat selfish since it will also help downstream, for a voidcrew TG rebase.

## Changelog

:cl:
refactor: Shuttle engines now handle engine stuff themselves, rather than delegating it to the shuttle, which would periodically scan all their tiles for engines.
fix: Cargo/Arrival/Pirate shuttles are now admin-replacable, and won't stack on top of eachother.
/:cl:

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
why is it doing this to me

**Comment by Addust** [comment]
you have my approval for this shuttle addition, especially if more engines make the shuttle faster as such a thing would hold actual logic

**Comment by LemonInTheDark** [comment]
> you have my approval for this shuttle addition, especially if more engines make the shuttle faster as such a thing would hold actual logic

We actually already do this, or should

**Comment by JohnFulpWillard** [comment]
> you have my approval for this shuttle addition

wow thanks

> especially if more engines make the shuttle faster as such a thing would hold actual logic

this is already the case, Im making engines add/remove themselves from shuttles, rather than shuttles constantly checking their contents for engines.

File: code/modules/shuttle/shuttle.dm [hunk 1]
```diff
@@ -436,54 +453,48 @@
...
 
-// Called after the shuttle is loaded from template
-/obj/docking_port/mobile/proc/linkup(obj/docking_port/stationary/dock)
-	for(var/place in shuttle_areas)
-		var/area/area = place
-		area.connect_to_shuttle(src, dock)
-		for(var/each in place)
-			var/atom/atom = each
-			atom.connect_to_shuttle(src, dock)
+	return INITIALIZE_HINT_LATELOAD //we need to load engines first
```
Comment (SuperNovaa41) [inline code/modules/shuttle/shuttle.dm:490]: is there a `LateInitialize()` that im not seeing?

File: code/modules/shuttle/shuttle.dm [hunk 2]
```diff
@@ -846,18 +855,17 @@
...
+	if(!istype(dockedAt, /obj/docking_port/stationary/transit))
 		return docked_at
+	if(timeLeft() > 1 HOURS)
+		return "Hyperspace"
+	else
+		var/obj/docking_port/stationary/dst
+		if(mode == SHUTTLE_RECALL)
+			dst = previous
+		else
+			dst = destination
```
Comment (SuperNovaa41) [inline code/modules/shuttle/shuttle.dm:867]:
```suggestion
		var/obj/docking_port/stationary/dst = (mode == SHUTTLE_RECALL) ? previous : destination
```

**Review by SuperNovaa41 — CHANGES_REQUESTED** [review]

File: code/modules/shuttle/shuttle.dm [hunk 3]
```diff
@@ -436,54 +453,46 @@
...
 /obj/docking_port/mobile/Destroy(force)
-	if(force)
-		unregister()
-		destination = null
-		previous = null
-		QDEL_NULL(assigned_transit) //don't need it where we're goin'!
-		shuttle_areas = null
-		remove_ripples()
-	. = ..()
+	unregister()
```
Comment (SuperNovaa41) [inline code/modules/shuttle/shuttle.dm:456]: why did you remove the `force` check here?

Comment (JohnFulpWillard) [reply code/modules/shuttle/shuttle.dm:456]: If we're deleting it, we should clear everything, force or not. I would rather we fix any problem where it's somehow deleted without force where we don't want it to be, instead of keeping this snowflake check.
